File: maximum_length_of_a_concatenated_string_with_unique_characters.go
func solve(arr []string, index int, mp []int) int {
    if index == len(arr) {
        return 0
    }

    var take int
    notTake := solve(arr, index+1, mp)
    flag := false

    for _, ch := range arr[index] {
        if mp[ch-'a'] == 1 {
            flag = true
        }
        mp[ch-'a']++
    }

    if flag {
        take = -1e9
    } else {
        take = len(arr[index]) + solve(arr, index+1, mp)
    }

    for _, ch := range arr[index] {
        mp[ch-'a']--
    }
    return max(take, notTake)
}

func maxLength(arr []string) int {
    mp := make([]int, 26)
    return solve(arr, 0, mp)
}

// Time Complexity: O(2^n)
// Space Complexity: O(2^n)

File: maximum_length_of_a_concatenated_string_with_unique_characters.cpp
class Solution {
public:
    int solve(vector<string>& arr, int index, vector<int>& mp) {
        if(index == arr.size()) return 0;

        bool flag = false;
        int take, notTake = solve(arr, index+1, mp);

        for(auto ch: arr[index]) {
            if(mp[ch-'a'] == 1) flag = true;
            mp[ch-'a']++;
        }
        if(flag) take = INT_MIN;
        else take = arr[index].size() + solve(arr, index+1, mp);

        for(auto ch: arr[index]) {
            mp[ch-'a']--;
        }
        
        return max(take, notTake);
    }
    int maxLength(vector<string>& arr) {
        vector<int> mp(26, 0);
        return solve(arr, 0, mp);
    }
};

// Time Complexity : O(2^n)
// Space Complexity : O(2^n)
